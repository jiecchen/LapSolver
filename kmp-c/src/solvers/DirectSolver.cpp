#include "DirectSolver.h"
#include "mkl.h"

template<MatrixType M>
struct DirectSolver<M>::DSS_impl
{
    MKL_INT opt = MKL_DSS_DEFAULTS;
    MKL_INT sym;
    MKL_INT type;
    _MKL_DSS_HANDLE_t handle;
    _INTEGER_t err = MKL_DSS_SUCCESS;
};

constexpr MKL_INT isSym(MatrixType mt)
{
    return (mt == Real || mt == RealPosDef)
           ? MKL_DSS_NON_SYMMETRIC
           : MKL_DSS_SYMMETRIC;
}

constexpr MKL_INT defType(MatrixType mt)
{
    return (mt == Real)
           ? MKL_DSS_INDEFINITE
           : ((mt == RealPosDef)
              ? MKL_DSS_POSITIVE_DEFINITE
              : ((mt == RealSym)
                 ? MKL_DSS_HERMITIAN_INDEFINITE
                 : MKL_DSS_HERMITIAN_POSITIVE_DEFINITE));
}

#define dss_guard(f, ...) \
    _impl->err = f(_impl->handle, __VA_ARGS__); \
    if (_impl->err != MKL_DSS_SUCCESS) \
        throw fprintf(stderr, "Error! " #f " failed with code %d\n", _impl->err), _impl->err;

template<MatrixType M>
DirectSolver<M>::DirectSolver(CSRMatrix<M> *csr)
    : csr(csr),
      dim(csr->getDimension())
{
    _impl = std::unique_ptr<DSS_impl>();
    _impl->sym = isSym(M);
    _impl->type = defType(M);

    dss_guard(dss_create, _impl->opt);
    dss_guard(dss_define_structure, _impl->sym,
              csr->rows.data(), csr->dim, csr->dim,
              csr->cols.data(), csr->nnz);
    dss_guard(dss_reorder, _impl->opt, NULL);
    dss_guard(dss_factor_real, _impl->type, csr->data.data());
}

template<MatrixType M>
DirectSolver<M>::~DirectSolver()
{
    dss_guard(dss_delete, _impl->opt);
}

template<MatrixType M>
void DirectSolver<M>::apply(double *b, double *x)
{
    dss_guard(dss_solve_real, _impl->opt, b, 1, x);
}
