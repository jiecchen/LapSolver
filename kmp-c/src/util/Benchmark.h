#pragma once
#include <string>
#include <chrono>
#include <stdexcept>
#include <cstdio>

template<typename Func>
class Benchmark
{
public:
    Benchmark(int argc, char **argv, Func f)
        : name(argv[0])
    {
        for (int i = 1; i < argc; ++i)
        {
            std::string flag(argv[i]);
            if (flag == "-v")
                echo = true;
            else if (flag == "-n" && i + 1 < argc)
            {
                try
                {
                    nTrials = std::stoi(argv[i + 1]);
                }
                catch (...)
                {
                    fprintf(stderr, "bad value for flag -n '%s'\n", argv[i + 1]);
                    break;
                }
            }
        }
        runBenchmark(f);
    }

    Benchmark(std::string name, Func f, bool echo = false, int nTrials = 1)
        : name(name), echo(echo), nTrials(nTrials)
    {
        runBenchmark(f);
    }

    ~Benchmark()
    {
        if (echo)
            fprintf(stderr, "%s completed in %dms (on average over %d trials)\n", name.c_str(), avgTime.count(), nTrials);
    }

    void runBenchmark(Func f)
    {
        auto start = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < nTrials; ++i)
            f();
        auto stop = std::chrono::high_resolution_clock::now();
        avgTime = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start) / nTrials;
    }

private:
    std::string name;
    bool echo = false;
    int nTrials = 1;
    std::chrono::milliseconds avgTime;
};

template<typename Func>
Benchmark<Func> make_benchmark(int argc, char **argv, Func f)
{
    return Benchmark<Func>(argc, argv, f);
}
