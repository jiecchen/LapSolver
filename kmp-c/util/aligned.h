#pragma once
#include "aligned_allocator.h"
#include "aligned_types.h"
#include "aligned_opers.h"
