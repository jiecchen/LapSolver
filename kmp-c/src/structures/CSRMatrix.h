#pragma once
#include "util/aligned.h"
#include "solvers/VectorOperator.h"
#include "mkl.h"

struct CSRMatrix : public VectorOperator
{
    using VectorOperator::apply;
    virtual void apply(double *x, double *y)
    {
        char n = 'n';
        mkl_cspblas_dcsrgemv(&n, &dim, data.data(), cols.data(), rows.data(), x, y);
    }

    virtual int getDimension() const
    {
        return dim;
    }

    aligned_vector<double> data;
    aligned_vector<int> cols;
    aligned_vector<int> rows;

    int dim;
    int nnz;

};


