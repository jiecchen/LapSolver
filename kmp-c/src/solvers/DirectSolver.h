#pragma once
#include <memory>
#include "structures/CSRMatrix.h"
#include "VectorOperator.h"
#include "mkl.h"

#define dss_guard(f, ...) \
    err = f(handle, ## __VA_ARGS__); \
    if (err != MKL_DSS_SUCCESS) \
        throw fprintf(stderr, "Error! " #f " failed with code %d\n", err), err;

template <MatrixType M>
class DirectSolver : public VectorOperator
{
public:
    using VectorOperator::apply;
    DirectSolver(CSRMatrix<M> *csr)
        : csr(csr),
          dim(csr->getDimension())
    {
        sym = (M == Real || M == RealPosDef)
              ? MKL_DSS_NON_SYMMETRIC
              : MKL_DSS_SYMMETRIC;
        type = (M == Real || M == RealSym)
               ? MKL_DSS_INDEFINITE
               : MKL_DSS_POSITIVE_DEFINITE;

        opt = MKL_DSS_DEFAULTS + MKL_DSS_ZERO_BASED_INDEXING;
        dss_guard(dss_create, opt);
        opt = MKL_DSS_DEFAULTS;
        dss_guard(dss_define_structure, sym,
                  csr->rows.data(), csr->dim, csr->dim,
                  csr->cols.data(), csr->nnz);
        dss_guard(dss_reorder, opt, NULL);
        dss_guard(dss_factor_real, type, csr->data.data());
    }

    virtual ~DirectSolver()
    {
        dss_guard(dss_delete, opt);
    }

    virtual int getDimension() const
    {
        return dim;
    }

    virtual void apply(double *b, double *x)
    {
        MKL_INT nRhs = 1;
        dss_guard(dss_solve_real, opt, b, nRhs, x);
    }

private:
    int dim;
    CSRMatrix<M> *csr;

    MKL_INT opt;
    MKL_INT sym;
    MKL_INT type;
    _MKL_DSS_HANDLE_t handle;
    _INTEGER_t err;
};
