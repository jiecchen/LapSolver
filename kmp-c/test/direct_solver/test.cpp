#include <solvers/DirectSolver.h>
#include <util/aligned.h>
#include <util/Benchmark.h>
#include <structures/CSRMatrix.h>
#include "mkl.h"

int main(int argc, char **argv)
{
    atexit(mkl_free_buffers);

    aligned_vector<int> ia = { 1, 6, 7, 8, 9, 10 };
    aligned_vector<int> ja = { 1, 2, 3, 4, 5, 2, 3, 4, 5 };
    aligned_vector<double> da = { 9, 1.5, 6, .75, 3, 0.5, 12, .625, 16 };

    // Change to 0-based indexing
    ia.data()[0:ia.size()]--;
    ja.data()[0:ja.size()]--;

    CSRMatrix<RealSym> a(ia, ja, da);
    DirectSolver<RealSym> dss(&a);

    // initialize solution via original operator
    aligned_vector<double> rhs = { 1, 2, 3, 4, 5 };

    aligned_vector<double> x(8);
    auto bench = make_benchmark(argc, argv, [&] ()
    {
        x = dss.apply(rhs);
    });

    for (auto xi : x)
        printf("%.3f\n", xi);

    return 0;
}
