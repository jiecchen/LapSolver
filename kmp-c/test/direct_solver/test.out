-326.333
983.000
163.417
398.000
61.500
